css-transitions-2/Overview.bs

```diff
@@ -167,6 +167,114 @@ Each time a new transition is generated, the current value of the (already
 incremented) <a>current transition generation</a> is stored as the
 transition's <dfn>transition generation</dfn>.
 
+## Defining [=before-change style=]: the ''@starting-style'' rule
+
+In Level 1 of this specification, transitions can only start during a
+[=style change event=] for elements which have a defined [=before-change style=]
+established by the previous [=style change event=]. That means a transition
+could not be started on an element that was not being rendered for the previous
+[=style change event=] (see: [[css-transitions-1#starting]]).
+
+In some cases it makes sense to start transitions on newly inserted elements or
+elements that change from not [=being rendered=] to being rendered. To allow
+for that, this specification introduces ''@starting-style''.
+
+The <dfn at-rule id="at-ruledef-starting-style">@starting-style</dfn> rule is a
+grouping rule where the rules inside it apply if the previous
+[=style change event=] did not establish a [=before-change style=] for the
+element whose styles are being computed. This is used to establish a
+[=before-change style=] for the current [=style change event=].
+
+Note: This means that ''@starting-style'' rules only apply to some elements
+during a computed style update, namely elements that were not rendered or part
+of the DOM during the previous [=style change event=].
+
+If any style rules in ''@starting-style'' are applied, the resulting style must
+be used as the [=before-change style=] for the current [=style change event=].
```

Comment:
My intent is that #b should not transition color for that change. As I mentioned in another comment, that you disagreed with, my intent is that starting style for an element inherits from the after-change style of the parent, regardless of whether any @starting-style rules applied to any of the ancestors.

Perhaps that's not consistent with how before-change style behaves, but as mentioned in another comment, the main reason I want to do it that way is that otherwise we need to keep two full style trees during style recalc and create separate @starting-style based computed style  instances for all elements (that did not already have a before-change style) in case there is some arbitrary descendant that can start a transition.

Reply:
Correction: we don't need to keep the whole tree alive, but the child needs to know what the before-change style of the parent is. That's currently generally lost in the current Blink implementation, apart from some container queries cases, but it still means that we need to compute two sets of computed styles for subtrees of an element that didn't have a before-change style and matches any @starting-style rules making it different than the after-change style.

The worst case for that is:

```css
html {
  color: orange;
  @starting-style {
    color: pink;
  }
}

#some-descendant {
  transition: color 1s;
}
```

We then need to compute two sets of computed styles for every element on the initial rendering in case there is a descendant that relies on the root element's starting style.

Reply:
@dbaron If I read you correctly, I think where we do not agree on the intent is about inheritance. I don't think that is covered by discussions and resolutions in the working group, so it probably makes sense to formulate this in a new github issue and involve the working group to get a resolution on that. Does that sound good to you?

Reply:
I think your comment at https://github.com/w3c/csswg-drafts/pull/8876#discussion_r1236644072 is convincing enough that I'm OK with your current approach to inheritance, assuming it's specified clearly enough (see my comments at the start of this thread at https://github.com/w3c/csswg-drafts/pull/8876#discussion_r1235303098 ).

Reply:
I've tried to rework the spec text now.